core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,589 @@
+    decode = D.decodeBytes
+
+instance Bi Text.Text where
+    encode = E.encodeString
+    decode = D.decodeString
+
+instance Bi BS.Lazy.ByteString where
+    encode = encode . BS.Lazy.toStrict
+    decode = BS.Lazy.fromStrict <$> decode
+
+instance Bi a => Bi [a] where
+    encode = encodeList
+    decode = decodeList
+
+instance (Bi a, Bi b) => Bi (Either a b) where
+    encode (Left  x) = E.encodeListLen 2 <> E.encodeWord 0 <> encode x
+    encode (Right x) = E.encodeListLen 2 <> E.encodeWord 1 <> encode x
+
+    decode = do D.decodeListLenOf 2
+                t <- D.decodeWord
+                case t of
+                  0 -> do !x <- decode
+                          return (Left x)
+                  1 -> do !x <- decode
+                          return (Right x)
+                  _ -> fail $ "decode@Either: unknown tag " <> show t
+
+instance Bi a => Bi (NonEmpty a) where
+  encode = defaultEncodeList . toList
+  decode = do
+    l <- defaultDecodeList
+    case nonEmpty l of
+      Nothing -> fail "Expected a NonEmpty list, but an empty list was found!"
+      Just xs -> return xs
+
+instance Bi a => Bi (Maybe a) where
+    encode Nothing  = E.encodeListLen 0
+    encode (Just x) = E.encodeListLen 1 <> encode x
+
+    decode = do n <- D.decodeListLen
+                case n of
+                  0 -> return Nothing
+                  1 -> do !x <- decode
+                          return (Just x)
+                  _ -> fail $ "decode@Maybe: unknown tag " <> show n
+
+encodeContainerSkel :: (Word -> E.Encoding)
+                    -> (container -> Int)
+                    -> (accumFunc -> E.Encoding -> container -> E.Encoding)
+                    -> accumFunc
+                    -> container
+                    -> E.Encoding
+encodeContainerSkel encodeLen size foldr f  c =
+    encodeLen (fromIntegral (size c)) <> foldr f mempty c
+{-# INLINE encodeContainerSkel #-}
+
+decodeContainerSkelWithReplicate
+  :: Bi a
+  => D.Decoder s Int
+     -- ^ How to get the size of the container
+  -> (Int -> D.Decoder s a -> D.Decoder s container)
+     -- ^ replicateM for the container
+  -> ([container] -> container)
+     -- ^ concat for the container
+  -> D.Decoder s container
+decodeContainerSkelWithReplicate decodeLen replicateFun fromList = do
+    -- Look at how much data we have at the moment and use it as the limit for
+    -- the size of a single call to replicateFun. We don't want to use
+    -- replicateFun directly on the result of decodeLen since this might lead to
+    -- DOS attack (attacker providing a huge value for length). So if it's above
+    -- our limit, we'll do manual chunking and then combine the containers into
+    -- one.
+    size <- decodeLen
+    limit <- D.peekAvailable
+    if size <= limit
+       then replicateFun size decode
+       else do
+           -- Take the max of limit and a fixed chunk size (note: limit can be
+           -- 0). This basically means that the attacker can make us allocate a
```

Comment:
BTW, this code is copied from the serialise package and was previously reviewed there by me and others.

I believe it's fine. What it does is look at how much data is remaining in the buffer (which is what the user has already gone to the effort of supplying to us) and use that as the chunk size. We then read in chunks of that size. That means we actually consume input, so the maximum we are allocating up front is the chunk size. That size is at smallest 128, which isn't much, and at most it's the amount of data the attacker has already supplied to us up front (ie they've already done the work).

With the example of maxBound as a length prefix (ie limit) then it means we'll start reading chunks, but the attacker has to actually supply us with 4Gb of input data to produce this large list of chunks. So that's ok for this layer. The top level layer deals with the overall message size. This layer just has to make sure it does not pre-allocate lots of memory, that it's pre-allocation is strictly limited and that we're only forced to allocate as much as the attacker actually supplies (+/- a constant).